From 5ccd517d13c9ecd2a2054480be6ee30bf53281fd Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Thu, 15 Jan 2015 19:42:02 +0000
Subject: [PATCH] Reboot: allow reboot as only action

When a standalone server is powered down it may only be rebooted
by manual intervention. With this patch LINUX_REBOOT_CMD_RESTART
is the only command accepted by reboot syscall depending on
configuration.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 kernel/power/Kconfig | 12 ++++++++++++
 kernel/reboot.c      |  5 +++++
 2 files changed, 17 insertions(+)

diff --git a/kernel/power/Kconfig b/kernel/power/Kconfig
index 48b28d3..7feed0c 100644
--- a/kernel/power/Kconfig
+++ b/kernel/power/Kconfig
@@ -300,3 +300,15 @@ config PM_GENERIC_DOMAINS_OF
 
 config CPU_PM
 	bool
+
+config REBOOT_ONLY
+	bool "Only allow reboot"
+	default n
+	help
+	  On stand alone servers a power down may be unwise as the
+	  system cannot be booted without manual intervention.
+
+	  If this flag is set, the reboot syscall will execute a
+	  reboot when requested to halt or power down.
+
+	  If in doubt, say N.
diff --git a/kernel/reboot.c b/kernel/reboot.c
index 5925f5a..7c7bec9 100644
--- a/kernel/reboot.c
+++ b/kernel/reboot.c
@@ -296,6 +296,11 @@ SYSCALL_DEFINE4(reboot, int, magic1, int, magic2, unsigned int, cmd,
 			magic2 != LINUX_REBOOT_MAGIC2C))
 		return -EINVAL;
 
+#if defined CONFIG_REBOOT_ONLY
+	if (cmd == LINUX_REBOOT_CMD_POWER_OFF || cmd == LINUX_REBOOT_CMD_HALT)
+		cmd = LINUX_REBOOT_CMD_RESTART;
+#endif
+
 	/*
 	 * If pid namespaces are enabled and the current task is in a child
 	 * pid_namespace, the command is handled by reboot_pid_ns() which will
-- 
2.1.4

